### PR TITLE
[typing] update `prefect.futures`

### PR DESCRIPTION
**Pull request by zzstoatzz**
related to #16292

updates typing of wait and `DoneAndNotDoneFutures`

### CODE REVIEW

**Review by desertaxle — APPROVED** [review]
LGTM!

File: src/prefect/futures.py [hunk 1]
```diff
@@ -331,21 +334,22 @@ def result(
...
+    futures: list[PrefectFuture[R]], timeout: Optional[float] = None
 ) -> Generator[PrefectFuture[R], None]:
-    unique_futures: Set[PrefectFuture[R]] = set(futures)
+    unique_futures: set[PrefectFuture[R]] = set(futures)
     total_futures = len(unique_futures)
+    pending = unique_futures
     try:
         with timeout_context(timeout):
-            done = {f for f in unique_futures if f._final_state}
+            done = {f for f in unique_futures if f._final_state}  # type: ignore[privateUsage]
```
Comment (mjpieters) [inline src/prefect/futures.py:344]: Perhaps the `PrefectFuture` class should have a public read-only property or method for this? Also, I find `getattr()` helpful to hide private access from the type checker 😇

File: src/prefect/futures.py [hunk 2]
```diff
@@ -431,9 +444,11 @@ def resolve_futures_to_states(
 
     Unsupported object types will be returned without modification.
     """
-    futures: Set[PrefectFuture[R]] = set()
+    futures: set[PrefectFuture[R]] = set()
 
-    def _collect_futures(futures, expr, context):
+    def _collect_futures(
+        futures: set[PrefectFuture[R]], expr: Any, context: Any
```
Comment (mjpieters) [inline src/prefect/futures.py:450]: `context` must at least support `.get(str)` so I think it should be a `Mapping[str, Any]`, certainly given what I've seen of how contexts are used in perfect. Ditto in other argument annotations below.

File: src/prefect/futures.py [hunk 3]
```diff
@@ -455,14 +470,14 @@ def _collect_futures(futures, expr, context):
         return expr
 
     # Get final states for each future
-    states = []
+    states: list[State] = []
```
Comment (mjpieters) [inline src/prefect/futures.py:473]: State is generic, so this is missing a parameter.

```suggestion
    states: list[State[Any]] = []
```

**Pull request merged**